### DISCUSSION TITLE
Performance implications of calling an C/C++ Extension's method

### DISCUSSION THREAD

**Discussion by szabolcsdombi** [pyodide/pyodide#4178] [General]
I am porting multiple projects to the web using Pyodide and I am interested in improving the performance when calling a C/C++ Extension's method.

Previously, I had made several calls to C/C++ Extensions for intensive operations.
However, recently I began using PyGLM, where the Python code handles numerous small objects like `vec3`, `vec4`, and `quat`.
In native Python installations, thousands of calls to an Extension per frame pose no issues.
However, my web port appears to take significantly longer with these calls.

I learned from [this podcast](https://realpython.com/podcasts/rpp/154/) that Emscripten can provide dynamic linking with trampolining.
If I understand correctly, calling a method from an Extension module is invoking JavaScript, which then calls another wasm module, with the results likely being returned in a similar manner.

One potential solution I'm considering is to build Pyodide using a custom subset of core modules, which would include my own. If this is done, I presume that the interpreter's calls to my extensions would eliminate the need for a JS roundtrip.

Questions:

1. Does calling a simple C/C++ Extension method require such trampolining?
2. What's the recommended method for modifying the core modules bundled with Pyodide?

**Comment by ryanking13** [comment]
> Does calling a simple C/C++ Extension method require such trampolining?

Yes, calling functions from a side module that is dynamically linked to the main module requires invoking JavaScript, and it might make things slower.

> What's the recommended method for modifying the core modules bundled with Pyodide?

There is no standard or recommended way to do it at least for now. I think you need to patch Pyodide makefile to include the library you built when making [pyodide.asm.js, pyodide.asm.wasm](https://github.com/pyodide/pyodide/blob/47024f5dacbbe596963125b210211dbc1340f834/Makefile#L90).

**Comment by hoodmane** [comment]
You should be aware though that `call_indirect` is not very much faster than a js trampoline.

**Reply by szabolcsdombi** [reply]
I am not sure what `call_indirect` is. I tried search for it.
My code usually operates on objecs like the builtin bytes, unicode, list, ...., except they come from an extension and do different things.
Here is an example:

```py
for obj in game_objects:
  obj.position += obj.velocity
  obj.velocity += obj.acceleration
  obj.acceleration *= 0.0
  obj.rotation = ...
  # and so on
```

do I use it in such a case?

PS: I invoke my per frame function like:

```js
const updateCallback = pyodide.globals.get('render');

const render = () => {
    updateCallback(); // 60 calls / sec, no lookup by name
    requestAnimationFrame(render);
};
requestAnimationFrame(render);
```
